{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents \n",
    "1. [Introduction](#table-of-contents)\n",
    "    - 1.1 [Project Description](#introduction)\n",
    "    - 1.2 [Imports](#imports)\n",
    "    - 1.3 [Data Dictionary](#data-dictionary)\n",
    "2. [Exploratory Data Analysis](#investigation-of-features)\n",
    "3. [Data Transformation](#data-transformation)\n",
    "    - 3.1 [Target Variable](#1-target-variable-cancer-type)\n",
    "    - 3.2 [Genomic Features](#2-genonic-features)\n",
    "    - 3.3 [Patient Features](#3-patient-features)\n",
    "4. [Feature Engineering](#feature-engineering)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "The primary goal of this project is to determine the genetic and lifestyle characteristics that are most predictive of brain tumor type (between Low Grade Glioma and Glioblastoma Multiforme)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GOALS\n",
    "* Two types of brain cancer, Glioma (GMA) and Glioblasoma (GBM).\n",
    "Build a classificatin model that distinguishes beetween two types bsed on genomic data. Eg variant classification, tumor sequance, allele, model that distinguishes one type or another. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import mean_squared_error  \n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import roc_auc_score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Dictionary\n",
    "\n",
    "#### All Initial Features\n",
    "| Field | Datatype | Description |\n",
    "| --- | --- | --- |\n",
    "| chrom | object | Chromosome\n",
    "| chromStart | int64 | Start position in chromosome\n",
    "| chromEnd | int64 | End position in chromosome\n",
    "| name | object | Name of item\n",
    "| score | int64 | Score from 0-1000\n",
    "| strand | object | + or -\n",
    "| thickStart | int64 | Start of where display should be thick (start codon)\n",
    "| thickEnd | int64 | End of where display should be thick (stop codon)\n",
    "| reserved | object | Used as itemRgb as of 2004-11-22\n",
    "| blockCount | int64 | Number of blocks\n",
    "| blockSizes | int64 | Comma separated list of block sizes\n",
    "| chromStarts | int64 | Start positions relative to chromStart\n",
    "| sampleCount | int64 | Number of samples with this variant\n",
    "| freq | float64 | Variant frequency\n",
    "| Hugo_Symbol | object | Symbol / nomenclature to identify each gene \n",
    "| Entrez_Gene_Id | int64 | Entrez Gene Id\n",
    "| Variant_Classification | object | Class of variant\n",
    "| Variant_Type | object | Type of variant\n",
    "| Reference_Allele | object | Reference allele\n",
    "| Tumor_Seq_Allele1 | object | Tumor allele 1\n",
    "| Tumor_Seq_Allele2 | object | Tumor allele 2\n",
    "| dbSNP_RS | object | dbSNP RS number\n",
    "| dbSNP_Val_Status | object | dbSNP validation status\n",
    "| days_to_death | object | Number of days till death\n",
    "| cigarettes_per_day | object | Number of cigarettes per day\n",
    "| weight | object | Weight\n",
    "| alcohol_history | object | Any alcohol consumption?\n",
    "| alcohol_intensity | object | Frequency of alcohol consumption\n",
    "| bmi | object | Body mass index\n",
    "| years_smoked | object | Number of years smoked\n",
    "| height | object | Height\n",
    "| gender | object | Gender\n",
    "| project_id | object | TCGA Project id\n",
    "| ethnicity | object | Ethnicity\n",
    "| Tumor_Sample_Barcode | object |  Tumor sample barcode\n",
    "| Matched_Norm_Sample_Barcode | object | Matched normal sample barcode\n",
    "| case_id | object | Case ID \n",
    "| cancer_type | object | Low Grade Glioma (LGG) / Gliomablastoma (GBM)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Dropped Columns\n",
    "* 'reserved'\n",
    "    - There is only one value for this feature\n",
    "* 'Tumor_Sample_Barcode'\n",
    "    - This feature is not relevant for modelling\n",
    "* 'Matched_Norm_Sample_Barcode'\n",
    "    - This feature is not relevant for modelling\n",
    "* 'case_id'\n",
    "    - This feature is not relevant for modelling\n",
    "* 'project_id'\n",
    "    - This feature is not relevant for modelling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Included Features\n",
    "\n",
    "#### Genomic Location Information \n",
    "| Feature | Description |\n",
    "| --- | --- |\n",
    "| chrom | Chromosome where the mutation is located (e.g. \"chr1\") |\n",
    "| chromStart & chromEnd | Precise position of the mutation on hte chromosome |\n",
    "| name | Simple description of the mutation (e.g., \"C>T\" means a C base was changed to a T) |\n",
    "| strand | DNA strand information (indicated by \".\") |\n",
    "\n",
    "#### Mutation Details\n",
    "| Feature | Description |\n",
    "| --- | --- |\n",
    "| Hugo_Symbol | Gene name affected by this mutation (e.g., \"OR4F5\", \"MIR6723\"). The Hugo symbol is the standard nomenclature for genes. |\n",
    "| Variant_Classification | Type of mutation effect (e.g., \"Silent\" means it doesn't change the protein) |\n",
    "| Variant_Type | Type of genetic change (SNP = Single Nucleotide Polymorphism) |\n",
    "| Reference_Allele | The original DNA base at this position | \n",
    "| Tumor_Seq_Allele1 & Tumor_Seq_Allele2 | The observed DNA bases in the tumor |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df = pd.read_csv('data/somatic_variant_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(118321, 38)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#\"chrom\"</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>reserved</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>...</th>\n",
       "      <th>bmi</th>\n",
       "      <th>years_smoked</th>\n",
       "      <th>height</th>\n",
       "      <th>gender</th>\n",
       "      <th>project_id</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>case_id</th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>TCGA-19-5956-01A-11D-1696-08</td>\n",
       "      <td>TCGA-19-5956-10A-01D-1696-08</td>\n",
       "      <td>7e70a318-3acc-42f3-8287-158a2199f0bc</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>male</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>TCGA-06-1800-01A-01W-0643-08</td>\n",
       "      <td>TCGA-06-1800-10A-01W-0643-08</td>\n",
       "      <td>b1ce95bf-7441-4aac-8131-19a177d2b8ac</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>TCGA-DU-6392-01A-11D-1705-08</td>\n",
       "      <td>TCGA-DU-6392-10A-01D-1705-08</td>\n",
       "      <td>fcd9e1c4-bddb-4856-844c-03df48fba499</td>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>TCGA-06-5416-01A-01D-1486-08</td>\n",
       "      <td>TCGA-06-5416-10A-01D-1486-08</td>\n",
       "      <td>3a3fc890-1985-4353-861b-dc3abfb364b1</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>male</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>TCGA-HW-7490-01A-11D-2024-08</td>\n",
       "      <td>TCGA-HW-7490-10A-01D-2024-08</td>\n",
       "      <td>c3c41fd2-f0d7-4f46-9e18-cce4626538c2</td>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #\"chrom\"  chromStart  chromEnd name  score strand  thickStart  thickEnd  \\\n",
       "0     chr1       69665     69666  C>T      1      .       69665     69666   \n",
       "1     chr1      634047    634048  G>A      1      .      634047    634048   \n",
       "2     chr1      874213    874214  C>T      1      .      874213    874214   \n",
       "3     chr1      874359    874360  T>G      1      .      874359    874360   \n",
       "4     chr1      925927    925928  G>C      1      .      925927    925928   \n",
       "\n",
       "  reserved  blockCount  ...  bmi  years_smoked  height  gender project_id  \\\n",
       "0    0,0,0           1  ...   --            --      --  female   TCGA-GBM   \n",
       "1    0,0,0           1  ...   --            --      --    male   TCGA-GBM   \n",
       "2    0,0,0           1  ...   --            --      --  female   TCGA-LGG   \n",
       "3    0,0,0           1  ...   --            --      --  female   TCGA-GBM   \n",
       "4    0,0,0           1  ...   --            --      --    male   TCGA-LGG   \n",
       "\n",
       "                ethnicity          Tumor_Sample_Barcode  \\\n",
       "0  not hispanic or latino  TCGA-19-5956-01A-11D-1696-08   \n",
       "1  not hispanic or latino  TCGA-06-1800-01A-01W-0643-08   \n",
       "2  not hispanic or latino  TCGA-DU-6392-01A-11D-1705-08   \n",
       "3  not hispanic or latino  TCGA-06-5416-01A-01D-1486-08   \n",
       "4  not hispanic or latino  TCGA-HW-7490-01A-11D-2024-08   \n",
       "\n",
       "    Matched_Norm_Sample_Barcode                               case_id  \\\n",
       "0  TCGA-19-5956-10A-01D-1696-08  7e70a318-3acc-42f3-8287-158a2199f0bc   \n",
       "1  TCGA-06-1800-10A-01W-0643-08  b1ce95bf-7441-4aac-8131-19a177d2b8ac   \n",
       "2  TCGA-DU-6392-10A-01D-1705-08  fcd9e1c4-bddb-4856-844c-03df48fba499   \n",
       "3  TCGA-06-5416-10A-01D-1486-08  3a3fc890-1985-4353-861b-dc3abfb364b1   \n",
       "4  TCGA-HW-7490-10A-01D-2024-08  c3c41fd2-f0d7-4f46-9e18-cce4626538c2   \n",
       "\n",
       "  cancer_type  \n",
       "0         GBM  \n",
       "1         GBM  \n",
       "2         LGG  \n",
       "3         GBM  \n",
       "4         LGG  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#\"chrom\"                           0\n",
       "chromStart                         0\n",
       "chromEnd                           0\n",
       "name                               0\n",
       "score                              0\n",
       "strand                             0\n",
       "thickStart                         0\n",
       "thickEnd                           0\n",
       "reserved                           0\n",
       "blockCount                         0\n",
       "blockSizes                         0\n",
       "chromStarts                        0\n",
       "sampleCount                        0\n",
       "freq                               0\n",
       "Hugo_Symbol                        0\n",
       "Entrez_Gene_Id                     0\n",
       "Variant_Classification             0\n",
       "Variant_Type                       0\n",
       "Reference_Allele                   0\n",
       "Tumor_Seq_Allele1                  0\n",
       "Tumor_Seq_Allele2                  0\n",
       "dbSNP_RS                       20171\n",
       "dbSNP_Val_Status               97248\n",
       "days_to_death                      0\n",
       "cigarettes_per_day                 0\n",
       "weight                             0\n",
       "alcohol_history                    0\n",
       "alcohol_intensity                  0\n",
       "bmi                                0\n",
       "years_smoked                       0\n",
       "height                             0\n",
       "gender                             0\n",
       "project_id                         0\n",
       "ethnicity                          0\n",
       "Tumor_Sample_Barcode               0\n",
       "Matched_Norm_Sample_Barcode        0\n",
       "case_id                            0\n",
       "cancer_type                        0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count all the missing values in the DataFrame\n",
    "cancer_df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  dbSNP_RS dbSNP_Val_Status\n",
       "0    novel              NaN\n",
       "1      NaN              NaN\n",
       "2    novel              NaN\n",
       "3    novel              NaN\n",
       "4    novel              NaN"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at the two features with missing values\n",
    "cancer_df[['dbSNP_RS', 'dbSNP_Val_Status']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dbSNP_RS\n",
       "novel          70425\n",
       "rs121913500      374\n",
       "rs121913343       50\n",
       "rs121913499       27\n",
       "rs139236063       17\n",
       "               ...  \n",
       "rs780611504        1\n",
       "rs778051530        1\n",
       "rs142559400        1\n",
       "rs77327288         1\n",
       "rs777620757        1\n",
       "Name: count, Length: 26461, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df['dbSNP_RS'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dbSNP_Val_Status\n",
       "byFrequency                                             7667\n",
       "byCluster;byFrequency                                   3838\n",
       "byCluster                                               2721\n",
       "by1000G;byCluster                                       2526\n",
       "by1000G;byCluster;byFrequency                           2347\n",
       "by1000G                                                 1449\n",
       "by1000G;byFrequency                                      414\n",
       "by1000G;byCluster;byFrequency;byHapMap                    40\n",
       "byCluster;byFrequency;byHapMap                            18\n",
       "byCluster;byHapMap                                        14\n",
       "byHapMap                                                   7\n",
       "by1000G;byCluster;byFrequency;suspect                      7\n",
       "by1000G;byCluster;byHapMap                                 5\n",
       "byCluster;suspect                                          5\n",
       "by1000G;byCluster;suspect                                  3\n",
       "suspect                                                    2\n",
       "byCluster;byFrequency;suspect                              2\n",
       "by1000G;suspect                                            1\n",
       "byHapMap;suspect                                           1\n",
       "by2Hit2Allele                                              1\n",
       "by1000G;byCluster;byHapMap;suspect                         1\n",
       "by1000G;byFrequency;suspect                                1\n",
       "byFrequency;suspect                                        1\n",
       "byCluster;byFrequency;byOtherPop;suspect                   1\n",
       "by1000G;by2Hit2Allele;byCluster;byFrequency;byHapMap       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df['dbSNP_Val_Status'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although these rows have missing values, they are not useful for our analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#\"chrom\"                       0\n",
       "chromStart                     0\n",
       "chromEnd                       0\n",
       "name                           0\n",
       "score                          0\n",
       "strand                         0\n",
       "thickStart                     0\n",
       "thickEnd                       0\n",
       "reserved                       0\n",
       "blockCount                     0\n",
       "blockSizes                     0\n",
       "chromStarts                    0\n",
       "sampleCount                    0\n",
       "freq                           0\n",
       "Hugo_Symbol                    0\n",
       "Entrez_Gene_Id                 0\n",
       "Variant_Classification         0\n",
       "Variant_Type                   0\n",
       "Reference_Allele               0\n",
       "Tumor_Seq_Allele1              0\n",
       "Tumor_Seq_Allele2              0\n",
       "days_to_death                  0\n",
       "cigarettes_per_day             0\n",
       "weight                         0\n",
       "alcohol_history                0\n",
       "alcohol_intensity              0\n",
       "bmi                            0\n",
       "years_smoked                   0\n",
       "height                         0\n",
       "gender                         0\n",
       "project_id                     0\n",
       "ethnicity                      0\n",
       "Tumor_Sample_Barcode           0\n",
       "Matched_Norm_Sample_Barcode    0\n",
       "case_id                        0\n",
       "cancer_type                    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1 = cancer_df.drop(['dbSNP_RS', 'dbSNP_Val_Status'], axis=1)\n",
    "cancer_df1.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The features with missing values have been dropped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for duplicates\n",
    "cancer_df1.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         1\n",
       "1         1\n",
       "2         1\n",
       "3         1\n",
       "4         1\n",
       "         ..\n",
       "118316    1\n",
       "118317    1\n",
       "118318    1\n",
       "118319    1\n",
       "118320    1\n",
       "Name: score, Length: 118321, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df['score']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are no duplicate rows found in this dataframe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following features are dropped because the do not afford data that is useful for modelling. They contain sample id info used for organization of the samples rather than for modelling. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop the following columns\n",
    "cancer_df1 = cancer_df1.drop(['case_id', 'Matched_Norm_Sample_Barcode', 'Tumor_Sample_Barcode'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Investigation of Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "name\n",
       "G>A                    28471\n",
       "C>T                    28093\n",
       "G>T                     9669\n",
       "C>A                     9657\n",
       "T>C                     6990\n",
       "                       ...  \n",
       "insAGAAAAACACAAGAAT        1\n",
       "insGACAGAGTGAGACTC         1\n",
       "insGATTGATG                1\n",
       "insAAG                     1\n",
       "insTCTATTTTTAT             1\n",
       "Name: count, Length: 1410, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "TTN             478\n",
       "TP53            429\n",
       "IDH1            414\n",
       "ATRX            244\n",
       "MUC16           176\n",
       "               ... \n",
       "DOHH              1\n",
       "RP11-38C17.1      1\n",
       "SDF4              1\n",
       "FAM132A           1\n",
       "UBE2J2            1\n",
       "Name: count, Length: 18113, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Hugo_Symbol'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Hugo symbol is nomenclature used to identify genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "reserved\n",
       "0,0,0    118321\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['reserved'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sincnne the 'reserved' column has only one value, we can drop it\n",
    "cancer_df1 = cancer_df1.drop('reserved', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Reference_Allele\n",
       "G                                     42458\n",
       "C                                     41970\n",
       "T                                     15292\n",
       "A                                     14380\n",
       "-                                      2879\n",
       "                                      ...  \n",
       "TCGCCTGTTAATTTTCTAAAAATAA                 1\n",
       "GAGCAGATGTTGGTAAAGAGTCAAAGGCAGGAGA        1\n",
       "CACGAAGGACTTGGTA                          1\n",
       "AAGATCTGTTTACGTGA                         1\n",
       "AGCTCGCTACTATAAA                          1\n",
       "Name: count, Length: 529, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Reference_Allele'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "529"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Reference_Allele'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18113"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Hugo_Symbol'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference Allele\n",
    "* The DNA base that is the standard sequence of a nucleotide (base) at a specific location in the reference genome (established sequence of DNA used to represent the standard version of the human genome). \n",
    "\n",
    "Hugo Symbol\n",
    "* Standard nomenclature used to identify genes. 18,113 of the 47,977 HUGO symbols are found in this dataset. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ordinal encoding? Creating a class system?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cancer_type\n",
       "GBM    82765\n",
       "LGG    35556\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['cancer_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: title={'center': 'Cancer Type Counts'}, xlabel='cancer_type'>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAHXCAYAAAC1evABAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8ekN5oAAAACXBIWXMAAA9hAAAPYQGoP6dpAABHSklEQVR4nO3de1hVdf73/9eOwxYQ9qAIuEc8VEQSloYNgpU2KnhAmruZUaMYHU0tUmKCTL9NZd0T5tlpmBpzHC0PQ3NPWZMmYSeLUdRIKjxVpokJYokbNNogrN8f/VzftpgJmujy+biudV3t9Xmvvd5rKfLqs9da22YYhiEAAAALuqy1GwAAAPipEHQAAIBlEXQAAIBlEXQAAIBlEXQAAIBlEXQAAIBlEXQAAIBlEXQAAIBlEXQAAIBlEXSAC9hHH32k3//+9+rWrZvatGmjtm3b6vrrr9esWbN0+PDh1m7vnNu7d69sNtsZLXv37m3tdj243W7l5ubqxhtvVHBwsHx9ffXzn/9cI0aM0Pr161u7PUnSgQMHNH36dJWUlLR2K8B5493aDQA4tUWLFik9PV1RUVF64IEHFB0drfr6er3//vv629/+po0bN2rVqlWt3eY51bFjR23cuNFjXXp6ulwul1asWNGk9kLx1VdfafDgwfroo480duxYPfDAA2rXrp2+/PJLvfLKKxowYICKi4t13XXXtWqfBw4c0GOPPaauXbuqZ8+erdoLcL4QdIAL0MaNG3XPPfdo0KBBevnll2W3282xQYMGKSsrS/n5+a3Y4dn75ptv5O/v77HObrerT58+HuuCgoJUV1fXZP2F5He/+50+/PBDvf766/rlL3/pMTZq1Cjdf//9Cg4ObqXugEsbH10BF6CcnBzZbDY9++yzHiHnBF9fX6WkpJivX3jhBSUmJqpjx47y8/NT9+7dNXXqVB07dsxjuzFjxqht27b67LPPNHToULVt21YRERHKysqS2+32qHW73Xr88cfVvXt3tWnTRu3bt9ctt9yiDRs2mDWGYejpp59Wz5495efnp+DgYP3mN7/R559/7vFe/fv3V0xMjN59910lJCTI399fY8eObdG5GTBggK6++mqd/H3EhmHoyiuv1LBhwyT978dgs2bN0hNPPKHOnTurTZs26t27t958880m7/vpp58qNTVVoaGhstvt6t69u/7617/+aD/FxcVau3atxo0b1yTknHDDDTeoc+fO5uvS0lLdeuutCg4OVps2bdSzZ08999xzHtssXbr0lB/RvfPOO7LZbHrnnXfMdSfO75YtW3TTTTfJ399fl19+uZ588kk1Njaa291www2SpN///vfmR4DTp0+XJH3++ecaNWqUnE6n7Ha7wsLCNGDAAD7mwkWPoANcYBoaGvTWW28pNjZWERERZ7TNp59+qqFDh2rx4sXKz89XZmam/vWvf2n48OFNauvr65WSkqIBAwbolVde0dixYzV//nzNnDnTrDl+/LiGDBmi//t//6+Sk5O1atUqLV26VAkJCdq3b59ZN3HiRGVmZmrgwIF6+eWX9fTTT2vbtm1KSEjQwYMHPfZbXl6uO++8U6mpqXrttdeUnp7eovNz3333adeuXU3Cytq1a7V7927de++9Hutzc3OVn5+vBQsWaPny5brssss0ZMgQj4/Itm/frhtuuEGlpaWaO3euVq9erWHDhikjI0OPPfbYafspKCiQJP3qV786o/537dqlhIQEbdu2TU899ZReeuklRUdHa8yYMZo1a9YZvcepVFRU6I477tCdd96p//znPxoyZIimTZum5cuXS5Kuv/56LVmyRJL0xz/+URs3btTGjRt11113SZKGDh2q4uJizZo1S+vWrdMzzzyjXr166ciRIy3uCbggGAAuKBUVFYYkY9SoUS3avrGx0aivrzfWr19vSDI+/PBDc2z06NGGJONf//qXxzZDhw41oqKizNfPP/+8IclYtGjRD+5n48aNhiRj7ty5HuvLysoMPz8/Y8qUKea6fv36GZKMN998s9nH069fP+Oaa64xXzc0NBiXX365ceutt3rUDRkyxLjiiiuMxsZGwzAMY8+ePYYkw+l0GrW1tWZddXW10a5dO2PgwIHmuqSkJKNTp06Gy+XyeM9JkyYZbdq0MQ4fPvyD/d19992GJGPnzp1ndDyjRo0y7Ha7sW/fvib9+/v7G0eOHDEMwzCWLFliSDL27NnjUff2228bkoy3337bXHfi/G7atMmjNjo62khKSjJfb9myxZBkLFmyxKPuq6++MiQZCxYsOKNjAC4mzOgAFvD5558rNTVV4eHh8vLyko+Pj/r16ydJ2rFjh0etzWZrMtNz7bXX6osvvjBfr127Vm3atDntx0urV6+WzWbTnXfeqePHj5tLeHi4rrvuOo+PViQpODj4Bz/aaY7LLrtMkyZN0urVq83Zpd27dys/P1/p6emy2Wwe9bfddpvatGljvg4MDNTw4cP17rvvqqGhQd9++63efPNN/Z//83/k7+/vcSxDhw7Vt99+q6KiorPu+4S33npLAwYMaDJbN2bMGH3zzTdNLsY+U+Hh4frFL37hse7kP9cf0q5dO11xxRWaPXu25s2bp61bt5ofeQEXO4IOcIEJCQmRv7+/9uzZc0b1R48e1U033aRNmzbpT3/6k9555x1t2bJFL730kiSptrbWo97f39/jF7/03UXA3377rfn60KFDcjqduuyyH/4n4uDBgzIMQ2FhYfLx8fFYioqK9NVXX3nUn8u7pMaOHSs/Pz/97W9/kyT99a9/lZ+f3ymDWXh4+CnX1dXV6ejRo/r66691/Phx/eUvf2lyHEOHDpWkJsfyfSeuvTnTP6+vv/76lOfC6XSa4y3Rvn37JuvsdnuTP/9TsdlsevPNN5WUlKRZs2bp+uuvV4cOHZSRkaGampoW9QNcKLjrCrjAeHl5acCAAVq7dq3279+vTp06nbb+rbfe0oEDB/TOO++YsziSzuraig4dOqiwsFCNjY0/GHZCQkJks9n03nvvnfKC6ZPXnTzTcjYcDodGjx6tv//978rOztaSJUuUmpqqn/3sZ01qKyoqTrnO19dXbdu2lY+Pj7y8vJSWltbk+p4TunXr9oO9JCUl6X/+53/08ssva/DgwT/ae/v27VVeXt5k/YEDByR9d14lmWH05IvETxe6zkaXLl20ePFiSdInn3yif/3rX5o+fbrq6urMQAlcjJjRAS5A06ZNk2EYGj9+vOrq6pqM19fX69VXX5X0vwHi5GCxcOHCFu9/yJAh+vbbb7V06dIfrElOTpZhGPryyy/Vu3fvJkuPHj1avP8zkZGRoa+++kq/+c1vdOTIEU2aNOmUdS+99JLHbFVNTY1effVV3XTTTfLy8pK/v79uueUWbd26Vddee+0pj+VUsyUnXH/99RoyZIgWL16st95665Q177//vvkx24ABA8xw+n3PP/+8/P39zdvou3btKum7h0Z+33/+85/Tn5jTOPF35Mdmea666ir98Y9/VI8ePfTBBx+0eH/AhYAZHeACFB8fr2eeeUbp6emKjY3VPffco2uuuUb19fXaunWrnn32WcXExGj48OFKSEhQcHCw7r77bj366KPy8fHRihUr9OGHH7Z4/7fffruWLFmiu+++W7t27dItt9yixsZGbdq0Sd27d9eoUaPUt29fTZgwQb///e/1/vvv6+abb1ZAQIDKy8tVWFioHj166J577jmHZ8XTVVddpcGDB2vt2rW68cYbf/BhfF5eXho0aJDuv/9+NTY2aubMmaqurva4m+rPf/6zbrzxRt10002655571LVrV9XU1Oizzz7Tq6+++oMB5oTnn39egwcP1pAhQzR27FgNGTJEwcHBKi8v16uvvqp//vOfKi4uVufOnfXoo49q9erVuuWWW/TII4+oXbt2WrFihdasWaNZs2bJ4XBI+u6W9KioKGVnZ+v48eMKDg7WqlWrVFhY2OJzdsUVV8jPz08rVqxQ9+7d1bZtWzmdTn311VeaNGmSfvvb3yoyMlK+vr5666239NFHH2nq1Kkt3h9wQWjli6EBnEZJSYkxevRoo3Pnzoavr68REBBg9OrVy3jkkUeMyspKs27Dhg1GfHy84e/vb3To0MG46667jA8++KDJHTajR482AgICmuzn0UcfNU7+56C2ttZ45JFHjMjISMPX19do37698ctf/tLYsGGDR90//vEPIy4uzggICDD8/PyMK664wvjd735nvP/++2bNyXdONcfptl26dKkhycjLy2syduKuq5kzZxqPPfaY0alTJ8PX19fo1auX8frrr5+yfuzYscbPf/5zw8fHx+jQoYORkJBg/OlPfzqjPmtra42nnnrKiI+PN4KCggxvb2/D6XQat912m7FmzRqP2o8//tgYPny44XA4DF9fX+O6665rcieUYRjGJ598YiQmJhpBQUFGhw4djMmTJxtr1qw55V1XpzpHo0ePNrp06eKx7p///Kdx9dVXGz4+PoYk49FHHzUOHjxojBkzxrj66quNgIAAo23btsa1115rzJ8/3zh+/PgZHT9wobIZxklP3QKAi8Svf/1rFRUVae/evfLx8fEY27t3r7p166bZs2crOzu7lToE0Nr46ArARcXtduuDDz7Q5s2btWrVKs2bN69JyAGAEwg6AC4q5eXlSkhIUFBQkCZOnKjJkye3dksALmB8dAUAACyL28sBAIBlEXQAAIBlEXQAAIBlXdIXIzc2NurAgQMKDAw8p4+nBwAAPx3DMFRTU/Oj38knXeJB58CBA02+QRgAAFwcysrKfvT7AC/poBMYGCjpuxMVFBTUyt0AAIAzUV1drYiICPP3+Olc0kHnxMdVQUFBBB0AAC4yZ3LZCRcjAwAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAyyLoAAAAy/Ju7QbQOrpOXdPaLeA82vvksNZuAQBaBTM6AADAsgg6AADAsgg6AADAsgg6AADAsgg6AADAsgg6AADAsgg6AADAspoVdI4fP64//vGP6tatm/z8/HT55Zfr8ccfV2Njo1ljGIamT58up9MpPz8/9e/fX9u2bfN4H7fbrcmTJyskJEQBAQFKSUnR/v37PWqqqqqUlpYmh8Mhh8OhtLQ0HTlyxKNm3759Gj58uAICAhQSEqKMjAzV1dU18xQAAACralbQmTlzpv72t78pNzdXO3bs0KxZszR79mz95S9/MWtmzZqlefPmKTc3V1u2bFF4eLgGDRqkmpoasyYzM1OrVq1SXl6eCgsLdfToUSUnJ6uhocGsSU1NVUlJifLz85Wfn6+SkhKlpaWZ4w0NDRo2bJiOHTumwsJC5eXl6cUXX1RWVtbZnA8AAGAhNsMwjDMtTk5OVlhYmBYvXmyu+/Wvfy1/f38tW7ZMhmHI6XQqMzNTDz74oKTvZm/CwsI0c+ZMTZw4US6XSx06dNCyZcs0cuRISdKBAwcUERGh1157TUlJSdqxY4eio6NVVFSkuLg4SVJRUZHi4+O1c+dORUVFae3atUpOTlZZWZmcTqckKS8vT2PGjFFlZaWCgoJ+9Hiqq6vlcDjkcrnOqN5KeDLypYUnIwOwkub8/m7WjM6NN96oN998U5988okk6cMPP1RhYaGGDh0qSdqzZ48qKiqUmJhobmO329WvXz9t2LBBklRcXKz6+nqPGqfTqZiYGLNm48aNcjgcZsiRpD59+sjhcHjUxMTEmCFHkpKSkuR2u1VcXHzK/t1ut6qrqz0WAABgXc36rqsHH3xQLpdLV199tby8vNTQ0KAnnnhCt99+uySpoqJCkhQWFuaxXVhYmL744guzxtfXV8HBwU1qTmxfUVGh0NDQJvsPDQ31qDl5P8HBwfL19TVrTjZjxgw99thjzTlkAABwEWvWjM4LL7yg5cuXa+XKlfrggw/03HPPac6cOXruuec86mw2m8drwzCarDvZyTWnqm9JzfdNmzZNLpfLXMrKyk7bEwAAuLg1a0bngQce0NSpUzVq1ChJUo8ePfTFF19oxowZGj16tMLDwyV9N9vSsWNHc7vKykpz9iU8PFx1dXWqqqrymNWprKxUQkKCWXPw4MEm+z906JDH+2zatMljvKqqSvX19U1mek6w2+2y2+3NOWQAAHARa9aMzjfffKPLLvPcxMvLy7y9vFu3bgoPD9e6devM8bq6Oq1fv94MMbGxsfLx8fGoKS8vV2lpqVkTHx8vl8ulzZs3mzWbNm2Sy+XyqCktLVV5eblZU1BQILvdrtjY2OYcFgAAsKhmzegMHz5cTzzxhDp37qxrrrlGW7du1bx58zR27FhJ332UlJmZqZycHEVGRioyMlI5OTny9/dXamqqJMnhcGjcuHHKyspS+/bt1a5dO2VnZ6tHjx4aOHCgJKl79+4aPHiwxo8fr4ULF0qSJkyYoOTkZEVFRUmSEhMTFR0drbS0NM2ePVuHDx9Wdna2xo8ff8ndQQUAAE6tWUHnL3/5ix5++GGlp6ersrJSTqdTEydO1COPPGLWTJkyRbW1tUpPT1dVVZXi4uJUUFCgwMBAs2b+/Pny9vbWiBEjVFtbqwEDBmjp0qXy8vIya1asWKGMjAzz7qyUlBTl5uaa415eXlqzZo3S09PVt29f+fn5KTU1VXPmzGnxyQAAANbSrOfoWA3P0cGlgufoALCSn+w5OgAAABcTgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALAsgg4AALCsZgWdrl27ymazNVnuvfdeSZJhGJo+fbqcTqf8/PzUv39/bdu2zeM93G63Jk+erJCQEAUEBCglJUX79+/3qKmqqlJaWpocDoccDofS0tJ05MgRj5p9+/Zp+PDhCggIUEhIiDIyMlRXV9eCUwAAAKyqWUFny5YtKi8vN5d169ZJkn77299KkmbNmqV58+YpNzdXW7ZsUXh4uAYNGqSamhrzPTIzM7Vq1Srl5eWpsLBQR48eVXJyshoaGsya1NRUlZSUKD8/X/n5+SopKVFaWpo53tDQoGHDhunYsWMqLCxUXl6eXnzxRWVlZZ3VyQAAANZiMwzDaOnGmZmZWr16tT799FNJktPpVGZmph588EFJ383ehIWFaebMmZo4caJcLpc6dOigZcuWaeTIkZKkAwcOKCIiQq+99pqSkpK0Y8cORUdHq6ioSHFxcZKkoqIixcfHa+fOnYqKitLatWuVnJyssrIyOZ1OSVJeXp7GjBmjyspKBQUFnVH/1dXVcjgccrlcZ7yNVXSduqa1W8B5tPfJYa3dAgCcM835/d3ia3Tq6uq0fPlyjR07VjabTXv27FFFRYUSExPNGrvdrn79+mnDhg2SpOLiYtXX13vUOJ1OxcTEmDUbN26Uw+EwQ44k9enTRw6Hw6MmJibGDDmSlJSUJLfbreLi4h/s2e12q7q62mMBAADW1eKg8/LLL+vIkSMaM2aMJKmiokKSFBYW5lEXFhZmjlVUVMjX11fBwcGnrQkNDW2yv9DQUI+ak/cTHBwsX19fs+ZUZsyYYV7343A4FBER0YwjBgAAF5sWB53FixdryJAhHrMqkmSz2TxeG4bRZN3JTq45VX1Lak42bdo0uVwucykrKzttXwAA4OLWoqDzxRdf6I033tBdd91lrgsPD5ekJjMqlZWV5uxLeHi46urqVFVVddqagwcPNtnnoUOHPGpO3k9VVZXq6+ubzPR8n91uV1BQkMcCAACsq0VBZ8mSJQoNDdWwYf97gWO3bt0UHh5u3oklfXcdz/r165WQkCBJio2NlY+Pj0dNeXm5SktLzZr4+Hi5XC5t3rzZrNm0aZNcLpdHTWlpqcrLy82agoIC2e12xcbGtuSQAACABXk3d4PGxkYtWbJEo0ePlrf3/25us9mUmZmpnJwcRUZGKjIyUjk5OfL391dqaqokyeFwaNy4ccrKylL79u3Vrl07ZWdnq0ePHho4cKAkqXv37ho8eLDGjx+vhQsXSpImTJig5ORkRUVFSZISExMVHR2ttLQ0zZ49W4cPH1Z2drbGjx/PLA0AADA1O+i88cYb2rdvn8aOHdtkbMqUKaqtrVV6erqqqqoUFxengoICBQYGmjXz58+Xt7e3RowYodraWg0YMEBLly6Vl5eXWbNixQplZGSYd2elpKQoNzfXHPfy8tKaNWuUnp6uvn37ys/PT6mpqZozZ05zDwcAAFjYWT1H52LHc3RwqeA5OgCs5Lw8RwcAAOBCR9ABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACWRdABAACW1eyg8+WXX+rOO+9U+/bt5e/vr549e6q4uNgcNwxD06dPl9PplJ+fn/r3769t27Z5vIfb7dbkyZMVEhKigIAApaSkaP/+/R41VVVVSktLk8PhkMPhUFpamo4cOeJRs2/fPg0fPlwBAQEKCQlRRkaG6urqmntIAADAopoVdKqqqtS3b1/5+Pho7dq12r59u+bOnauf/exnZs2sWbM0b9485ebmasuWLQoPD9egQYNUU1Nj1mRmZmrVqlXKy8tTYWGhjh49quTkZDU0NJg1qampKikpUX5+vvLz81VSUqK0tDRzvKGhQcOGDdOxY8dUWFiovLw8vfjii8rKyjqL0wEAAKzEZhiGcabFU6dO1X//+1+99957pxw3DENOp1OZmZl68MEHJX03exMWFqaZM2dq4sSJcrlc6tChg5YtW6aRI0dKkg4cOKCIiAi99tprSkpK0o4dOxQdHa2ioiLFxcVJkoqKihQfH6+dO3cqKipKa9euVXJyssrKyuR0OiVJeXl5GjNmjCorKxUUFPSjx1NdXS2HwyGXy3VG9VbSdeqa1m4B59HeJ4e1dgsAcM405/d3s2Z0/vOf/6h379767W9/q9DQUPXq1UuLFi0yx/fs2aOKigolJiaa6+x2u/r166cNGzZIkoqLi1VfX+9R43Q6FRMTY9Zs3LhRDofDDDmS1KdPHzkcDo+amJgYM+RIUlJSktxut8dHad/ndrtVXV3tsQAAAOtqVtD5/PPP9cwzzygyMlKvv/667r77bmVkZOj555+XJFVUVEiSwsLCPLYLCwszxyoqKuTr66vg4ODT1oSGhjbZf2hoqEfNyfsJDg6Wr6+vWXOyGTNmmNf8OBwORURENOfwAQDARaZZQaexsVHXX3+9cnJy1KtXL02cOFHjx4/XM88841Fns9k8XhuG0WTdyU6uOVV9S2q+b9q0aXK5XOZSVlZ22p4AAMDFrVlBp2PHjoqOjvZY1717d+3bt0+SFB4eLklNZlQqKyvN2Zfw8HDV1dWpqqrqtDUHDx5ssv9Dhw551Jy8n6qqKtXX1zeZ6TnBbrcrKCjIYwEAANbVrKDTt29f7dq1y2PdJ598oi5dukiSunXrpvDwcK1bt84cr6ur0/r165WQkCBJio2NlY+Pj0dNeXm5SktLzZr4+Hi5XC5t3rzZrNm0aZNcLpdHTWlpqcrLy82agoIC2e12xcbGNuewAACARXk3p/gPf/iDEhISlJOToxEjRmjz5s169tln9eyzz0r67qOkzMxM5eTkKDIyUpGRkcrJyZG/v79SU1MlSQ6HQ+PGjVNWVpbat2+vdu3aKTs7Wz169NDAgQMlfTdLNHjwYI0fP14LFy6UJE2YMEHJycmKioqSJCUmJio6OlppaWmaPXu2Dh8+rOzsbI0fP56ZGgAAIKmZQeeGG27QqlWrNG3aND3++OPq1q2bFixYoDvuuMOsmTJlimpra5Wenq6qqirFxcWpoKBAgYGBZs38+fPl7e2tESNGqLa2VgMGDNDSpUvl5eVl1qxYsUIZGRnm3VkpKSnKzc01x728vLRmzRqlp6erb9++8vPzU2pqqubMmdPikwEAAKylWc/RsRqeo4NLBc/RAWAlP9lzdAAAAC4mBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZzQo606dPl81m81jCw8PNccMwNH36dDmdTvn5+al///7atm2bx3u43W5NnjxZISEhCggIUEpKivbv3+9RU1VVpbS0NDkcDjkcDqWlpenIkSMeNfv27dPw4cMVEBCgkJAQZWRkqK6urpmHDwAArKzZMzrXXHONysvLzeXjjz82x2bNmqV58+YpNzdXW7ZsUXh4uAYNGqSamhqzJjMzU6tWrVJeXp4KCwt19OhRJScnq6GhwaxJTU1VSUmJ8vPzlZ+fr5KSEqWlpZnjDQ0NGjZsmI4dO6bCwkLl5eXpxRdfVFZWVkvPAwAAsCDvZm/g7e0xi3OCYRhasGCBHnroId12222SpOeee05hYWFauXKlJk6cKJfLpcWLF2vZsmUaOHCgJGn58uWKiIjQG2+8oaSkJO3YsUP5+fkqKipSXFycJGnRokWKj4/Xrl27FBUVpYKCAm3fvl1lZWVyOp2SpLlz52rMmDF64oknFBQU1OITAgAArKPZMzqffvqpnE6nunXrplGjRunzzz+XJO3Zs0cVFRVKTEw0a+12u/r166cNGzZIkoqLi1VfX+9R43Q6FRMTY9Zs3LhRDofDDDmS1KdPHzkcDo+amJgYM+RIUlJSktxut4qLi3+wd7fbrerqao8FAABYV7OCTlxcnJ5//nm9/vrrWrRokSoqKpSQkKCvv/5aFRUVkqSwsDCPbcLCwsyxiooK+fr6Kjg4+LQ1oaGhTfYdGhrqUXPyfoKDg+Xr62vWnMqMGTPM634cDociIiKac/gAAOAi06ygM2TIEP36179Wjx49NHDgQK1Zs0bSdx9RnWCz2Ty2MQyjybqTnVxzqvqW1Jxs2rRpcrlc5lJWVnbavgAAwMXtrG4vDwgIUI8ePfTpp5+a1+2cPKNSWVlpzr6Eh4errq5OVVVVp605ePBgk30dOnTIo+bk/VRVVam+vr7JTM/32e12BQUFeSwAAMC6zirouN1u7dixQx07dlS3bt0UHh6udevWmeN1dXVav369EhISJEmxsbHy8fHxqCkvL1dpaalZEx8fL5fLpc2bN5s1mzZtksvl8qgpLS1VeXm5WVNQUCC73a7Y2NizOSQAAGAhzbrrKjs7W8OHD1fnzp1VWVmpP/3pT6qurtbo0aNls9mUmZmpnJwcRUZGKjIyUjk5OfL391dqaqokyeFwaNy4ccrKylL79u3Vrl07ZWdnmx+FSVL37t01ePBgjR8/XgsXLpQkTZgwQcnJyYqKipIkJSYmKjo6WmlpaZo9e7YOHz6s7OxsjR8/nlkaAABgalbQ2b9/v26//XZ99dVX6tChg/r06aOioiJ16dJFkjRlyhTV1tYqPT1dVVVViouLU0FBgQIDA833mD9/vry9vTVixAjV1tZqwIABWrp0qby8vMyaFStWKCMjw7w7KyUlRbm5uea4l5eX1qxZo/T0dPXt21d+fn5KTU3VnDlzzupkAAAAa7EZhmG0dhOtpbq6Wg6HQy6X65KbCeo6dU1rt4DzaO+Tw1q7BQA4Z5rz+5vvugIAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZF0AEAAJZ1VkFnxowZstlsyszMNNcZhqHp06fL6XTKz89P/fv317Zt2zy2c7vdmjx5skJCQhQQEKCUlBTt37/fo6aqqkppaWlyOBxyOBxKS0vTkSNHPGr27dun4cOHKyAgQCEhIcrIyFBdXd3ZHBIAALCQFgedLVu26Nlnn9W1117rsX7WrFmaN2+ecnNztWXLFoWHh2vQoEGqqakxazIzM7Vq1Srl5eWpsLBQR48eVXJyshoaGsya1NRUlZSUKD8/X/n5+SopKVFaWpo53tDQoGHDhunYsWMqLCxUXl6eXnzxRWVlZbX0kAAAgMW0KOgcPXpUd9xxhxYtWqTg4GBzvWEYWrBggR566CHddtttiomJ0XPPPadvvvlGK1eulCS5XC4tXrxYc+fO1cCBA9WrVy8tX75cH3/8sd544w1J0o4dO5Sfn6+///3vio+PV3x8vBYtWqTVq1dr165dkqSCggJt375dy5cvV69evTRw4EDNnTtXixYtUnV19dmeFwAAYAEtCjr33nuvhg0bpoEDB3qs37NnjyoqKpSYmGius9vt6tevnzZs2CBJKi4uVn19vUeN0+lUTEyMWbNx40Y5HA7FxcWZNX369JHD4fCoiYmJkdPpNGuSkpLkdrtVXFx8yr7dbreqq6s9FgAAYF3ezd0gLy9PH3zwgbZs2dJkrKKiQpIUFhbmsT4sLExffPGFWePr6+sxE3Si5sT2FRUVCg0NbfL+oaGhHjUn7yc4OFi+vr5mzclmzJihxx577EwOEwAAWECzZnTKysp03333afny5WrTps0P1tlsNo/XhmE0WXeyk2tOVd+Smu+bNm2aXC6XuZSVlZ22JwAAcHFrVtApLi5WZWWlYmNj5e3tLW9vb61fv15PPfWUvL29zRmWk2dUKisrzbHw8HDV1dWpqqrqtDUHDx5ssv9Dhw551Jy8n6qqKtXX1zeZ6TnBbrcrKCjIYwEAANbVrKAzYMAAffzxxyopKTGX3r1764477lBJSYkuv/xyhYeHa926deY2dXV1Wr9+vRISEiRJsbGx8vHx8agpLy9XaWmpWRMfHy+Xy6XNmzebNZs2bZLL5fKoKS0tVXl5uVlTUFAgu92u2NjYFpwKAABgNc26RicwMFAxMTEe6wICAtS+fXtzfWZmpnJychQZGanIyEjl5OTI399fqampkiSHw6Fx48YpKytL7du3V7t27ZSdna0ePXqYFzd3795dgwcP1vjx47Vw4UJJ0oQJE5ScnKyoqChJUmJioqKjo5WWlqbZs2fr8OHDys7O1vjx45mpAQAAklpwMfKPmTJlimpra5Wenq6qqirFxcWpoKBAgYGBZs38+fPl7e2tESNGqLa2VgMGDNDSpUvl5eVl1qxYsUIZGRnm3VkpKSnKzc01x728vLRmzRqlp6erb9++8vPzU2pqqubMmXOuDwkAAFykbIZhGK3dRGuprq6Ww+GQy+W65GaBuk5d09ot4Dza++Sw1m4BAM6Z5vz+5ruuAACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZRF0AACAZZ3zby8HALQuvrT30sKX9p4eMzoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCymhV0nnnmGV177bUKCgpSUFCQ4uPjtXbtWnPcMAxNnz5dTqdTfn5+6t+/v7Zt2+bxHm63W5MnT1ZISIgCAgKUkpKi/fv3e9RUVVUpLS1NDodDDodDaWlpOnLkiEfNvn37NHz4cAUEBCgkJEQZGRmqq6tr5uEDAAAra1bQ6dSpk5588km9//77ev/99/XLX/5St956qxlmZs2apXnz5ik3N1dbtmxReHi4Bg0apJqaGvM9MjMztWrVKuXl5amwsFBHjx5VcnKyGhoazJrU1FSVlJQoPz9f+fn5KikpUVpamjne0NCgYcOG6dixYyosLFReXp5efPFFZWVlne35AAAAFmIzDMM4mzdo166dZs+erbFjx8rpdCozM1MPPvigpO9mb8LCwjRz5kxNnDhRLpdLHTp00LJlyzRy5EhJ0oEDBxQREaHXXntNSUlJ2rFjh6Kjo1VUVKS4uDhJUlFRkeLj47Vz505FRUVp7dq1Sk5OVllZmZxOpyQpLy9PY8aMUWVlpYKCgs6o9+rqajkcDrlcrjPexiq6Tl3T2i3gPNr75LDWbgHnET/fl5ZL8ee7Ob+/W3yNTkNDg/Ly8nTs2DHFx8drz549qqioUGJiolljt9vVr18/bdiwQZJUXFys+vp6jxqn06mYmBizZuPGjXI4HGbIkaQ+ffrI4XB41MTExJghR5KSkpLkdrtVXFz8gz273W5VV1d7LAAAwLqaHXQ+/vhjtW3bVna7XXfffbdWrVql6OhoVVRUSJLCwsI86sPCwsyxiooK+fr6Kjg4+LQ1oaGhTfYbGhrqUXPyfoKDg+Xr62vWnMqMGTPM634cDociIiKaefQAAOBi0uygExUVpZKSEhUVFemee+7R6NGjtX37dnPcZrN51BuG0WTdyU6uOVV9S2pONm3aNLlcLnMpKys7bV8AAODi1uyg4+vrqyuvvFK9e/fWjBkzdN111+nPf/6zwsPDJanJjEplZaU5+xIeHq66ujpVVVWdtubgwYNN9nvo0CGPmpP3U1VVpfr6+iYzPd9nt9vNO8ZOLAAAwLrO+jk6hmHI7XarW7duCg8P17p168yxuro6rV+/XgkJCZKk2NhY+fj4eNSUl5ertLTUrImPj5fL5dLmzZvNmk2bNsnlcnnUlJaWqry83KwpKCiQ3W5XbGzs2R4SAACwCO/mFP/P//yPhgwZooiICNXU1CgvL0/vvPOO8vPzZbPZlJmZqZycHEVGRioyMlI5OTny9/dXamqqJMnhcGjcuHHKyspS+/bt1a5dO2VnZ6tHjx4aOHCgJKl79+4aPHiwxo8fr4ULF0qSJkyYoOTkZEVFRUmSEhMTFR0drbS0NM2ePVuHDx9Wdna2xo8fzywNAAAwNSvoHDx4UGlpaSovL5fD4dC1116r/Px8DRo0SJI0ZcoU1dbWKj09XVVVVYqLi1NBQYECAwPN95g/f768vb01YsQI1dbWasCAAVq6dKm8vLzMmhUrVigjI8O8OyslJUW5ubnmuJeXl9asWaP09HT17dtXfn5+Sk1N1Zw5c87qZAAAAGs56+foXMx4jg4uFZficzYuZfx8X1ouxZ/v8/IcHQAAgAsdQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFgWQQcAAFhWs4LOjBkzdMMNNygwMFChoaH61a9+pV27dnnUGIah6dOny+l0ys/PT/3799e2bds8atxutyZPnqyQkBAFBAQoJSVF+/fv96ipqqpSWlqaHA6HHA6H0tLSdOTIEY+affv2afjw4QoICFBISIgyMjJUV1fXnEMCAAAW1qygs379et17770qKirSunXrdPz4cSUmJurYsWNmzaxZszRv3jzl5uZqy5YtCg8P16BBg1RTU2PWZGZmatWqVcrLy1NhYaGOHj2q5ORkNTQ0mDWpqakqKSlRfn6+8vPzVVJSorS0NHO8oaFBw4YN07Fjx1RYWKi8vDy9+OKLysrKOpvzAQAALMRmGIbR0o0PHTqk0NBQrV+/XjfffLMMw5DT6VRmZqYefPBBSd/N3oSFhWnmzJmaOHGiXC6XOnTooGXLlmnkyJGSpAMHDigiIkKvvfaakpKStGPHDkVHR6uoqEhxcXGSpKKiIsXHx2vnzp2KiorS2rVrlZycrLKyMjmdTklSXl6exowZo8rKSgUFBf1o/9XV1XI4HHK5XGdUbyVdp65p7RZwHu19clhrt4DziJ/vS8ul+PPdnN/fZ3WNjsvlkiS1a9dOkrRnzx5VVFQoMTHRrLHb7erXr582bNggSSouLlZ9fb1HjdPpVExMjFmzceNGORwOM+RIUp8+feRwODxqYmJizJAjSUlJSXK73SouLj5lv263W9XV1R4LAACwrhYHHcMwdP/99+vGG29UTEyMJKmiokKSFBYW5lEbFhZmjlVUVMjX11fBwcGnrQkNDW2yz9DQUI+ak/cTHBwsX19fs+ZkM2bMMK/5cTgcioiIaO5hAwCAi0iLg86kSZP00Ucf6Z///GeTMZvN5vHaMIwm6052cs2p6ltS833Tpk2Ty+Uyl7KystP2BAAALm4tCjqTJ0/Wf/7zH7399tvq1KmTuT48PFySmsyoVFZWmrMv4eHhqqurU1VV1WlrDh482GS/hw4d8qg5eT9VVVWqr69vMtNzgt1uV1BQkMcCAACsq1lBxzAMTZo0SS+99JLeeustdevWzWO8W7duCg8P17p168x1dXV1Wr9+vRISEiRJsbGx8vHx8agpLy9XaWmpWRMfHy+Xy6XNmzebNZs2bZLL5fKoKS0tVXl5uVlTUFAgu92u2NjY5hwWAACwKO/mFN97771auXKlXnnlFQUGBpozKg6HQ35+frLZbMrMzFROTo4iIyMVGRmpnJwc+fv7KzU11awdN26csrKy1L59e7Vr107Z2dnq0aOHBg4cKEnq3r27Bg8erPHjx2vhwoWSpAkTJig5OVlRUVGSpMTEREVHRystLU2zZ8/W4cOHlZ2drfHjxzNTAwAAJDUz6DzzzDOSpP79+3usX7JkicaMGSNJmjJlimpra5Wenq6qqirFxcWpoKBAgYGBZv38+fPl7e2tESNGqLa2VgMGDNDSpUvl5eVl1qxYsUIZGRnm3VkpKSnKzc01x728vLRmzRqlp6erb9++8vPzU2pqqubMmdOsEwAAAKzrrJ6jc7HjOTq4VFyKz9m4lPHzfWm5FH++z9tzdAAAAC5kBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZBB0AAGBZzQ467777roYPHy6n0ymbzaaXX37ZY9wwDE2fPl1Op1N+fn7q37+/tm3b5lHjdrs1efJkhYSEKCAgQCkpKdq/f79HTVVVldLS0uRwOORwOJSWlqYjR4541Ozbt0/Dhw9XQECAQkJClJGRobq6uuYeEgAAsKhmB51jx47puuuuU25u7inHZ82apXnz5ik3N1dbtmxReHi4Bg0apJqaGrMmMzNTq1atUl5engoLC3X06FElJyeroaHBrElNTVVJSYny8/OVn5+vkpISpaWlmeMNDQ0aNmyYjh07psLCQuXl5enFF19UVlZWcw8JAABYlHdzNxgyZIiGDBlyyjHDMLRgwQI99NBDuu222yRJzz33nMLCwrRy5UpNnDhRLpdLixcv1rJlyzRw4EBJ0vLlyxUREaE33nhDSUlJ2rFjh/Lz81VUVKS4uDhJ0qJFixQfH69du3YpKipKBQUF2r59u8rKyuR0OiVJc+fO1ZgxY/TEE08oKCioRScEAABYxzm9RmfPnj2qqKhQYmKiuc5ut6tfv37asGGDJKm4uFj19fUeNU6nUzExMWbNxo0b5XA4zJAjSX369JHD4fCoiYmJMUOOJCUlJcntdqu4uPiU/bndblVXV3ssAADAus5p0KmoqJAkhYWFeawPCwszxyoqKuTr66vg4ODT1oSGhjZ5/9DQUI+ak/cTHBwsX19fs+ZkM2bMMK/5cTgcioiIaMFRAgCAi8VPcteVzWbzeG0YRpN1Jzu55lT1Lan5vmnTpsnlcplLWVnZaXsCAAAXt3MadMLDwyWpyYxKZWWlOfsSHh6uuro6VVVVnbbm4MGDTd7/0KFDHjUn76eqqkr19fVNZnpOsNvtCgoK8lgAAIB1ndOg061bN4WHh2vdunXmurq6Oq1fv14JCQmSpNjYWPn4+HjUlJeXq7S01KyJj4+Xy+XS5s2bzZpNmzbJ5XJ51JSWlqq8vNysKSgokN1uV2xs7Lk8LAAAcJFq9l1XR48e1WeffWa+3rNnj0pKStSuXTt17txZmZmZysnJUWRkpCIjI5WTkyN/f3+lpqZKkhwOh8aNG6esrCy1b99e7dq1U3Z2tnr06GHehdW9e3cNHjxY48eP18KFCyVJEyZMUHJysqKioiRJiYmJio6OVlpammbPnq3Dhw8rOztb48ePZ6YGAABIakHQef/993XLLbeYr++//35J0ujRo7V06VJNmTJFtbW1Sk9PV1VVleLi4lRQUKDAwEBzm/nz58vb21sjRoxQbW2tBgwYoKVLl8rLy8usWbFihTIyMsy7s1JSUjye3ePl5aU1a9YoPT1dffv2lZ+fn1JTUzVnzpzmnwUAAGBJNsMwjNZuorVUV1fL4XDI5XJdcrNAXaeuae0WcB7tfXJYa7eA84if70vLpfjz3Zzf33zXFQAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsKyLPug8/fTT6tatm9q0aaPY2Fi99957rd0SAAC4QFzUQeeFF15QZmamHnroIW3dulU33XSThgwZon379rV2awAA4AJwUQedefPmady4cbrrrrvUvXt3LViwQBEREXrmmWdauzUAAHAB8G7tBlqqrq5OxcXFmjp1qsf6xMREbdiw4ZTbuN1uud1u87XL5ZIkVVdX/3SNXqAa3d+0dgs4jy7Fv+OXMn6+Ly2X4s/3iWM2DONHay/aoPPVV1+poaFBYWFhHuvDwsJUUVFxym1mzJihxx57rMn6iIiIn6RH4ELhWNDaHQD4qVzKP981NTVyOBynrblog84JNpvN47VhGE3WnTBt2jTdf//95uvGxkYdPnxY7du3/8FtYB3V1dWKiIhQWVmZgoKCWrsdAOcQP9+XFsMwVFNTI6fT+aO1F23QCQkJkZeXV5PZm8rKyiazPCfY7XbZ7XaPdT/72c9+qhZxgQoKCuIfQsCi+Pm+dPzYTM4JF+3FyL6+voqNjdW6des81q9bt04JCQmt1BUAALiQXLQzOpJ0//33Ky0tTb1791Z8fLyeffZZ7du3T3fffXdrtwYAAC4AF3XQGTlypL7++ms9/vjjKi8vV0xMjF577TV16dKltVvDBchut+vRRx9t8vElgIsfP9/4ITbjTO7NAgAAuAhdtNfoAAAA/BiCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsCyCDgAAsKyL+oGBwA/x8vI6o7qGhoafuBMAP7Wamhp9/5Fwl112mdq2bduKHeFCQtCBJRmGoS5dumj06NHq1atXa7cD4BwqKSnRQw89pDVr1kiSnE6nvvnmG3PcZrNp48aNuuGGG1qrRVxAeDIyLGnLli36xz/+oby8PHXr1k1jx47VHXfcoeDg4NZuDcBZGjdunK688kpNmzZNkhQYGKiFCxfq5z//uQzD0D/+8Q8ZhqFly5a1cqe4EBB0YGnffvut/v3vf2vJkiUqKirS8OHDNW7cOA0aNKi1WwPQQldffbUWLVqkm266SdJ3QefDDz/U5ZdfLknatGmTRowYoS+++KI128QFgouRYWlt2rTRnXfeqTfffFOlpaWqrKzU4MGDdfjw4dZuDUALlZWVqXPnzubrxx9/XCEhIebrjh076uDBg63RGi5AXKMDy9u/f7+WLl2qpUuXqra2Vg888ICCgoJauy0ALWS327V//3516dJFkvSHP/zBY7ysrEz+/v6t0RouQMzowJLq6ur0wgsvKDExUZGRkfrggw+0YMEClZWV6cknn5S3NxkfuFj16tVLL7/88g+Ov/TSS9yEABP/2sOSOnbsqMDAQI0ePVpPP/20QkNDJUlHjx71qGNmB7j4pKena9SoUeratavuueceXXbZd//P3tDQoKefflp/+ctftHLlylbuEhcKLkaGJZ34h0/67lbTkxmGIZvNxnN0gIvUgw8+qNmzZyswMFCXX365bDabdu/eraNHj+r+++/X7NmzW7tFXCAIOrCk9evXn1Fdv379fuJOAPxUioqK9M9//lOffvqpJCkyMlK33367+vTp08qd4UJC0AEAAJbFNTq4ZBiGobffflu1tbVKSEjg4YHARay6utq8xu61117T8ePHzTEvLy8NGzastVrDBYYZHVjSkSNHdN999+mDDz5Qnz59NHfuXA0dOlQbNmyQJHXo0EHr1q3Ttdde28qdAmiu1atX6+GHH9bWrVslfffAwGPHjpnjNptNL7zwgn7zm9+0Vou4gHB7OSwpOztbGzdu1MiRI/Xxxx9r8ODBamho0MaNG7Vp0yZFR0froYceau02AbTAs88+q0mTJnms++yzz9TY2KjGxkbNmDFD//jHP1qpO1xomNGBJf385z/XypUr1a9fP3355ZeKiIjQW2+9pf79+0uSNm/erJSUFFVUVLRuowCarWvXrvr3v/+t3r17S2r6FRAff/yxBgwYoMrKytZsExcIZnRgSQcPHtRVV10l6bvQ06ZNG0VERJjjnTt31qFDh1qrPQBnoaKiQu3btzdfv/322x4/323btpXL5WqN1nABIujAkhobG+Xl5WW+9vLy8niezqmerQPg4tCuXTvt3r3bfN27d2/5+PiYrz/99FO1a9euNVrDBYi7rmBZf//739W2bVtJ0vHjx7V06VLzi/9qampaszUAZ+Hmm2/WU089pYEDB55y/KmnntLNN998nrvChYprdGBJXbt2PaNZmz179pyHbgCcS1u3blV8fLyGDx+uKVOmmB9T79q1SzNnztSaNWu0YcMGXX/99a3cKS4EBB0AwEXnlVde0V133aXDhw97rA8ODtbf//53/epXv2qdxnDBIejAkr799lu98cYbSk5OliRNmzZNbrfbHPf29tbjjz+uNm3atFaLAM7SN998o9dff93jKyASExN1+PBhPfroo9xiDkkEHVjUwoULtXr1ar366quSvrv99JprrpGfn58kaefOnXrggQd0//33t2abAH4CH374oa6//nq+tBeSuOsKFrVixQqNHTvWY93KlSv19ttv6+2339bs2bP1//7f/2ul7gAA5wtBB5b0ySefmBcoSlKbNm102WX/+9f9F7/4hbZv394arQEAziNuL4cluVwueXv/71/vkx8O2NjY6HHNDgDAmgg6sKROnTqptLRUUVFRpxz/6KOP1KlTp/PcFYBz4bbbbjvt+JEjR85PI7goEHRgSUOHDtUjjzyiYcOGNbmzqra2Vo899piGDRvWSt0BOBsOh+NHx3/3u9+dp25woeOuK1jSwYMH1bNnT/n6+mrSpEm66qqrZLPZtHPnTuXm5ur48ePaunWrwsLCWrtVAMBPiKADy9qzZ4/uuecerVu3Tif+mttsNg0aNEhPP/20+U3HAADrIujA8g4fPqzPPvtMknTllVfyZX8AcAkh6AAAAMviOToAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoAAMCyCDoA8P/r2rWrFixY0NptADiHCDoALgl1dXWt3QKAVkDQAdAijY2Nmjlzpq688krZ7XZ17txZTzzxhCTpwQcf1FVXXSV/f39dfvnlevjhh1VfX29uO336dPXs2VPLli1T165d5XA4NGrUKNXU1JzR+0vSl19+qZEjRyo4OFjt27fXrbfeqr1795rjY8aM0a9+9SvNmDFDTqdTV1111WmPp3///vriiy/0hz/8QTabTTabTceOHVNQUJD+/e9/e9S++uqrCggIUE1Njfbu3Subzaa8vDwlJCSoTZs2uuaaa/TOO+94bLN9+3YNHTpUbdu2VVhYmNLS0vTVV18197QDaCaCDoAWmTZtmmbOnKmHH35Y27dv18qVK83vDgsMDNTSpUu1fft2/fnPf9aiRYs0f/58j+13796tl19+WatXr9bq1au1fv16Pfnkk2f0/t98841uueUWtW3bVu+++64KCwvVtm1bDR482GPm5s0339SOHTu0bt06rV69+rTH89JLL6lTp056/PHHVV5ervLycgUEBGjUqFFasmSJR+2SJUv0m9/8RoGBgea6Bx54QFlZWdq6dasSEhKUkpKir7/+WpJUXl6ufv36qWfPnnr//feVn5+vgwcPasSIES048wCaxQCAZqqurjbsdruxaNGiM6qfNWuWERsba75+9NFHDX9/f6O6utpc98ADDxhxcXFn9P6LFy82oqKijMbGRnOd2+02/Pz8jNdff90wDMMYPXq0ERYWZrjd7jM+ri5duhjz58/3WLdp0ybDy8vL+PLLLw3DMIxDhw4ZPj4+xjvvvGMYhmHs2bPHkGQ8+eST5jb19fVGp06djJkzZxqGYRgPP/ywkZiY6PG+ZWVlhiRj165dZ9wfgObzbu2gBeDis2PHDrndbg0YMOCU4//+97+1YMECffbZZzp69KiOHz+uoKAgj5quXbt6zIh07NhRlZWVZ/T+xcXF+uyzzzy2l6Rvv/1Wu3fvNl/36NFDvr6+LTrGE37xi1/ommuu0fPPP6+pU6dq2bJl6ty5s26++WaPuvj4ePO/vb291bt3b+3YscPs9+2331bbtm2bvP/u3bt/9GM1AC1H0AHQbH5+fj84VlRUpFGjRumxxx5TUlKSHA6H8vLyNHfuXI86Hx8fj9c2m02NjY0/+v7Sd9fvxMbGasWKFU3GOnToYP53QEDAjx7LmbjrrruUm5urqVOnasmSJfr9738vm832o9udqGlsbNTw4cM1c+bMJjUdO3Y8Jz0CODWu0QHQbJGRkfLz89Obb77ZZOy///2vunTpooceeki9e/dWZGSkvvjii3P2/pJ0/fXX69NPP1VoaKiuvPJKj8XhcLTomCTJ19dXDQ0NTdbfeeed2rdvn5566ilt27ZNo0ePblJTVFRk/vfx48dVXFysq6++2ux327Zt6tq1a5N+z1UYA3BqBB0AzdamTRs9+OCDmjJlip5//nnt3r1bRUVFWrx4sa688krt27dPeXl52r17t5566imtWrXqnL2/JN1xxx0KCQnRrbfeqvfee0979uzR+vXrdd9992n//v0tPq6uXbvq3Xff1ZdffulxR1RwcLBuu+02PfDAA0pMTFSnTp2abPvXv/5Vq1at0s6dO3XvvfeqqqpKY8eOlSTde++9Onz4sG6//XZt3rxZn3/+uQoKCjR27NhTBisA5w5BB0CLPPzww8rKytIjjzyi7t27a+TIkaqsrNStt96qP/zhD5o0aZJ69uypDRs26OGHHz5n7y9J/v7+evfdd9W5c2fddttt6t69u8aOHava2tom1wI1x+OPP669e/fqiiuu8PgITJLGjRunuro6M7yc7Mknn9TMmTN13XXX6b333tMrr7yikJAQSZLT6dR///tfNTQ0KCkpSTExMbrvvvvkcDh02WX8Mwz8lGyGYRit3QQAXOhWrFih++67TwcOHPC4wHnv3r3q1q2btm7dqp49e7ZegwBOiYuRAeA0vvnmG+3Zs0czZszQxIkTz/ouLgDnF3OmAC4J7733ntq2bfuDyw+ZNWuWevbsqbCwME2bNu08dgzgXOCjKwCXhNraWn355Zc/OH7llVeex24AnC8EHQAAYFl8dAUAACyLoAMAACyLoAMAACyLoAMAACyLoAMAACyLoAMAACyLoAMAACzr/wMYZ+gwM87wMQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cancer_df1['cancer_type'].value_counts().plot(kind='bar',\n",
    "                                                   title='Cancer Type Counts')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because there are only two values in the target variable, binary encoding will suffice. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant_Classification\n",
       "Missense_Mutation         68430\n",
       "Silent                    23974\n",
       "Nonsense_Mutation          5099\n",
       "3'UTR                      4839\n",
       "Intron                     4068\n",
       "Frame_Shift_Del            2384\n",
       "Frame_Shift_Ins            1887\n",
       "5'UTR                      1874\n",
       "RNA                        1646\n",
       "Splice_Site                1384\n",
       "Splice_Region              1061\n",
       "3'Flank                     500\n",
       "5'Flank                     466\n",
       "In_Frame_Del                293\n",
       "In_Frame_Ins                261\n",
       "Translation_Start_Site       85\n",
       "Nonstop_Mutation             64\n",
       "IGR                           6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df['Variant_Classification'].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 19 types of Variant Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant_Type\n",
       "SNP    112187\n",
       "DEL      3255\n",
       "INS      2879\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df['Variant_Type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* SNP (single nucleotide polymorphism): \n",
    "    - When one nucleotide (A, T, C, G) in a DNA sequence differs from a referene sequence. \n",
    "* DEL (deletion): \n",
    "    - Where a segment of DNA is lost or removed from its normal position in the genome.\n",
    "* INS (insertion): \n",
    "    - Where one or more nucleotides are added to a DNA sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop the 'strand' and 'reserved' columns as they have only one data value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df1 = cancer_df.drop(['strand',\n",
    "                            'reserved',\n",
    "                            'case_id', \n",
    "                            'Matched_Norm_Sample_Barcode', \n",
    "                            'Tumor_Sample_Barcode'], axis=1)\n",
    "                            \n",
    "                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#\"chrom\"</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>chromStarts</th>\n",
       "      <th>...</th>\n",
       "      <th>weight</th>\n",
       "      <th>alcohol_history</th>\n",
       "      <th>alcohol_intensity</th>\n",
       "      <th>bmi</th>\n",
       "      <th>years_smoked</th>\n",
       "      <th>height</th>\n",
       "      <th>gender</th>\n",
       "      <th>project_id</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>male</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>female</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>1</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>male</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #\"chrom\"  chromStart  chromEnd name  score  thickStart  thickEnd  \\\n",
       "0     chr1       69665     69666  C>T      1       69665     69666   \n",
       "1     chr1      634047    634048  G>A      1      634047    634048   \n",
       "2     chr1      874213    874214  C>T      1      874213    874214   \n",
       "3     chr1      874359    874360  T>G      1      874359    874360   \n",
       "4     chr1      925927    925928  G>C      1      925927    925928   \n",
       "\n",
       "   blockCount  blockSizes  chromStarts  ...  weight  alcohol_history  \\\n",
       "0           1           1            0  ...      --               --   \n",
       "1           1           1            0  ...      --               --   \n",
       "2           1           1            0  ...      --               --   \n",
       "3           1           1            0  ...      --               --   \n",
       "4           1           1            0  ...      --               --   \n",
       "\n",
       "  alcohol_intensity  bmi years_smoked height  gender project_id  \\\n",
       "0                --   --           --     --  female   TCGA-GBM   \n",
       "1                --   --           --     --    male   TCGA-GBM   \n",
       "2                --   --           --     --  female   TCGA-LGG   \n",
       "3                --   --           --     --  female   TCGA-GBM   \n",
       "4                --   --           --     --    male   TCGA-LGG   \n",
       "\n",
       "                ethnicity cancer_type  \n",
       "0  not hispanic or latino         GBM  \n",
       "1  not hispanic or latino         GBM  \n",
       "2  not hispanic or latino         LGG  \n",
       "3  not hispanic or latino         GBM  \n",
       "4  not hispanic or latino         LGG  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Transformation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Target Variable (Cancer Type)\n",
    " - Binary encode (LGG -> 0, GBM -> 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cancer_type\n",
       "GBM    82765\n",
       "LGG    35556\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['cancer_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binary encode the cancer_type column LGG = 0, GBM = 1\n",
    "cancer_df1['cancer_type'] = cancer_df1['cancer_type'].apply(lambda x: 0 if x == 'LGG' else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cancer_type\n",
       "1    82765\n",
       "0    35556\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to see if the encoding was successful\n",
    "cancer_df1['cancer_type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Genonic Features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 'chrom' \n",
    " - one-hot encode chromosomes (chr1, chr2, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the \"chrom\" column to name to chrom\n",
    "cancer_df1 = cancer_df1.rename(columns={'#\"chrom\"' : 'chrom'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# One hot encode the chromosome column\n",
    "cancer_df1 = pd.get_dummies(cancer_df1, columns=['chrom'],\n",
    "                            dtype=int,\n",
    "                            drop_first=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>chromStarts</th>\n",
       "      <th>sampleCount</th>\n",
       "      <th>...</th>\n",
       "      <th>chrom_chr22</th>\n",
       "      <th>chrom_chr3</th>\n",
       "      <th>chrom_chr4</th>\n",
       "      <th>chrom_chr5</th>\n",
       "      <th>chrom_chr6</th>\n",
       "      <th>chrom_chr7</th>\n",
       "      <th>chrom_chr8</th>\n",
       "      <th>chrom_chr9</th>\n",
       "      <th>chrom_chrX</th>\n",
       "      <th>chrom_chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>1</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromStart  chromEnd name  score  thickStart  thickEnd  blockCount  \\\n",
       "0       69665     69666  C>T      1       69665     69666           1   \n",
       "1      634047    634048  G>A      1      634047    634048           1   \n",
       "2      874213    874214  C>T      1      874213    874214           1   \n",
       "3      874359    874360  T>G      1      874359    874360           1   \n",
       "4      925927    925928  G>C      1      925927    925928           1   \n",
       "\n",
       "   blockSizes  chromStarts  sampleCount  ...  chrom_chr22 chrom_chr3  \\\n",
       "0           1            0            1  ...            0          0   \n",
       "1           1            0            1  ...            0          0   \n",
       "2           1            0            1  ...            0          0   \n",
       "3           1            0            1  ...            0          0   \n",
       "4           1            0            1  ...            0          0   \n",
       "\n",
       "   chrom_chr4 chrom_chr5 chrom_chr6 chrom_chr7 chrom_chr8 chrom_chr9  \\\n",
       "0           0          0          0          0          0          0   \n",
       "1           0          0          0          0          0          0   \n",
       "2           0          0          0          0          0          0   \n",
       "3           0          0          0          0          0          0   \n",
       "4           0          0          0          0          0          0   \n",
       "\n",
       "  chrom_chrX chrom_chrY  \n",
       "0          0          0  \n",
       "1          0          0  \n",
       "2          0          0  \n",
       "3          0          0  \n",
       "4          0          0  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check to see if the encoding was successful\n",
    "cancer_df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 'Variant_Classification'\n",
    " - one-hot encode (Silent, Missense, Nonsense, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant_Classification\n",
       "Missense_Mutation         68430\n",
       "Silent                    23974\n",
       "Nonsense_Mutation          5099\n",
       "3'UTR                      4839\n",
       "Intron                     4068\n",
       "Frame_Shift_Del            2384\n",
       "Frame_Shift_Ins            1887\n",
       "5'UTR                      1874\n",
       "RNA                        1646\n",
       "Splice_Site                1384\n",
       "Splice_Region              1061\n",
       "3'Flank                     500\n",
       "5'Flank                     466\n",
       "In_Frame_Del                293\n",
       "In_Frame_Ins                261\n",
       "Translation_Start_Site       85\n",
       "Nonstop_Mutation             64\n",
       "IGR                           6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Variant_Classification'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dummy variables for the 'Variant_Classification' column\n",
    "cancer_df1 = pd.get_dummies(cancer_df1, columns=['Variant_Classification'],\n",
    "                            dtype=int,\n",
    "                            drop_first=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>chromStarts</th>\n",
       "      <th>sampleCount</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant_Classification_In_Frame_Ins</th>\n",
       "      <th>Variant_Classification_Intron</th>\n",
       "      <th>Variant_Classification_Missense_Mutation</th>\n",
       "      <th>Variant_Classification_Nonsense_Mutation</th>\n",
       "      <th>Variant_Classification_Nonstop_Mutation</th>\n",
       "      <th>Variant_Classification_RNA</th>\n",
       "      <th>Variant_Classification_Silent</th>\n",
       "      <th>Variant_Classification_Splice_Region</th>\n",
       "      <th>Variant_Classification_Splice_Site</th>\n",
       "      <th>Variant_Classification_Translation_Start_Site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>69665</td>\n",
       "      <td>69666</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1</td>\n",
       "      <td>634047</td>\n",
       "      <td>634048</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>874213</td>\n",
       "      <td>874214</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>874359</td>\n",
       "      <td>874360</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>1</td>\n",
       "      <td>925927</td>\n",
       "      <td>925928</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 71 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromStart  chromEnd name  score  thickStart  thickEnd  blockCount  \\\n",
       "0       69665     69666  C>T      1       69665     69666           1   \n",
       "1      634047    634048  G>A      1      634047    634048           1   \n",
       "2      874213    874214  C>T      1      874213    874214           1   \n",
       "3      874359    874360  T>G      1      874359    874360           1   \n",
       "4      925927    925928  G>C      1      925927    925928           1   \n",
       "\n",
       "   blockSizes  chromStarts  sampleCount  ...  \\\n",
       "0           1            0            1  ...   \n",
       "1           1            0            1  ...   \n",
       "2           1            0            1  ...   \n",
       "3           1            0            1  ...   \n",
       "4           1            0            1  ...   \n",
       "\n",
       "   Variant_Classification_In_Frame_Ins Variant_Classification_Intron  \\\n",
       "0                                    0                             0   \n",
       "1                                    0                             0   \n",
       "2                                    0                             0   \n",
       "3                                    0                             0   \n",
       "4                                    0                             0   \n",
       "\n",
       "   Variant_Classification_Missense_Mutation  \\\n",
       "0                                         0   \n",
       "1                                         0   \n",
       "2                                         0   \n",
       "3                                         0   \n",
       "4                                         0   \n",
       "\n",
       "  Variant_Classification_Nonsense_Mutation  \\\n",
       "0                                        0   \n",
       "1                                        0   \n",
       "2                                        0   \n",
       "3                                        0   \n",
       "4                                        0   \n",
       "\n",
       "  Variant_Classification_Nonstop_Mutation Variant_Classification_RNA  \\\n",
       "0                                       0                          0   \n",
       "1                                       0                          0   \n",
       "2                                       0                          1   \n",
       "3                                       0                          1   \n",
       "4                                       0                          0   \n",
       "\n",
       "  Variant_Classification_Silent Variant_Classification_Splice_Region  \\\n",
       "0                             1                                    0   \n",
       "1                             0                                    0   \n",
       "2                             0                                    0   \n",
       "3                             0                                    0   \n",
       "4                             0                                    0   \n",
       "\n",
       "  Variant_Classification_Splice_Site  \\\n",
       "0                                  0   \n",
       "1                                  0   \n",
       "2                                  0   \n",
       "3                                  0   \n",
       "4                                  0   \n",
       "\n",
       "  Variant_Classification_Translation_Start_Site  \n",
       "0                                             0  \n",
       "1                                             0  \n",
       "2                                             0  \n",
       "3                                             0  \n",
       "4                                             0  \n",
       "\n",
       "[5 rows x 71 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check to see if the encoding was successful\n",
    "cancer_df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ' Variant_Type'\n",
    " - One-hot encode (SNP, INS, DEL, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant_Type\n",
       "SNP    112187\n",
       "DEL      3255\n",
       "INS      2879\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Variant_Type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dummy variables for the 'Variant_Type' feature\n",
    "cancer_df1 = pd.get_dummies(cancer_df1, columns=['Variant_Type'],\n",
    "                            dtype=int,\n",
    "                            drop_first=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "source": [
    "#### 'Hugo_Symbol'\n",
    " - Frequency encoding (how common each gene is)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "TTN             478\n",
       "TP53            429\n",
       "IDH1            414\n",
       "ATRX            244\n",
       "MUC16           176\n",
       "               ... \n",
       "DOHH              1\n",
       "RP11-38C17.1      1\n",
       "SDF4              1\n",
       "FAM132A           1\n",
       "UBE2J2            1\n",
       "Name: count, Length: 18113, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Hugo_Symbol'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since there are so many different unique Hugo_Symbol values we will drop the column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df1 = cancer_df1.drop('Hugo_Symbol', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Patient Features\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gender\n",
    " - Binary encod the 'gender' feature (female -> 0, male -> 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binary encode the 'gender' feature\n",
    "cancer_df1['gender'] = cancer_df1['gender'].apply(lambda x: 0 if x == 'female' else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gender\n",
       "0    70054\n",
       "1    48267\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['gender'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Engineering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "chromStart, chromEnd: Use as is, but consider:\n",
    " - Chromosome-normalized position (position + chromosome length)\n",
    " - Distance from centromere or telomere\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Genomic Context Features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation Length \n",
    "- mutation_length = chromEnd - chromStart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new feature called 'mutation_length'\n",
    "# which is the difference between the chrromEnd and chromStart columns\n",
    "\n",
    "cancer_df1['mutation_length'] = cancer_df1['chromEnd'] - cancer_df1['chromStart']\n",
    "cancer_df1['mutation_length'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exonic \n",
    " -  is_exonic (boolean): Flag if mutation is in a protein-coding region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "529"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Reference_Allele'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Family\n",
    " - Group genes by biological function/family"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromosome Group \n",
    " - Group chromosomes (sex chromosomes vs. autosomes)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Mutation Characteristics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transition vs Transversion\n",
    " - Flag for mutation type (transitions are A>G or C>T)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CpG Site\n",
    " - Flag if mutation occurs at a CpG site (often hypermutable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functional Impact\n",
    " - Combine Variant_Classification into severity groups\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_df1['Variant_Classification'] = cancer_df['Variant_Classification']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant_Classification\n",
       "Missense_Mutation         68430\n",
       "Silent                    23974\n",
       "Nonsense_Mutation          5099\n",
       "3'UTR                      4839\n",
       "Intron                     4068\n",
       "Frame_Shift_Del            2384\n",
       "Frame_Shift_Ins            1887\n",
       "5'UTR                      1874\n",
       "RNA                        1646\n",
       "Splice_Site                1384\n",
       "Splice_Region              1061\n",
       "3'Flank                     500\n",
       "5'Flank                     466\n",
       "In_Frame_Del                293\n",
       "In_Frame_Ins                261\n",
       "Translation_Start_Site       85\n",
       "Nonstop_Mutation             64\n",
       "IGR                           6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_df1['Variant_Classification'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Severe / Serious\n",
    "* Missense Mutation\n",
    "    - A type of point mutation\n",
    "* Silent\n",
    "    - \n",
    "\n",
    "#### \n",
    "\n",
    "#### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Advanced Genomic Features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Expression Cluster\n",
    " - If you have gene expression data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation Burden\n",
    " - mutation_burden: Count of mutations per patient\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway Impact\n",
    " - pathway_impact: Aggregate mutations by biological pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interaction Features\n",
    " - gender × gene: Some mutations have gender-specific effects "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modelling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Logistic Regression Model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Split the data\n",
    "    - Divide the dataset into training and test sets (80:20 ratio).\n",
    "2. Scaling parameters\n",
    "    - Standard Scaler will be used\n",
    "3. Scale the Training Data\n",
    "    - Apply the scaling transformation to the training data\n",
    "4. Fit the model\n",
    "    - Train the model using the scaled training data\n",
    "5. Scale the test data\n",
    "    - Scale the test data before making predictions. \n",
    "6. Evaluate model performance\n",
    "    - Using the scaled test data, evaluate the model's performance (accuracy, precision, recall, F1-score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Regularization: Lasso regularization will be used\n",
    "* Scaler: StandardScaler will be used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cancer_df1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# create a logistic regression model\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[43mcancer_df1\u001b[49m\u001b[38;5;241m.\u001b[39mdrop(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcancer_type\u001b[39m\u001b[38;5;124m'\u001b[39m, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      3\u001b[0m y \u001b[38;5;241m=\u001b[39m cancer_df1[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcancer_type\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cancer_df1' is not defined"
     ]
    }
   ],
   "source": [
    "# create a logistic regression model\n",
    "X = cancer_df1.drop('cancer_type', axis=1)\n",
    "y = cancer_df1['cancer_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Train-test split\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m X_train, X_test, y_train, y_test \u001b[38;5;241m=\u001b[39m train_test_split(\u001b[43mX\u001b[49m, y, test_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m, random_state\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m42\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X' is not defined"
     ]
    }
   ],
   "source": [
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X_train' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Instantiate the model\u001b[39;00m\n\u001b[1;32m      2\u001b[0m model \u001b[38;5;241m=\u001b[39m LogisticRegression()\n\u001b[0;32m----> 3\u001b[0m model\u001b[38;5;241m.\u001b[39mfit(\u001b[43mX_train\u001b[49m, y_train)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X_train' is not defined"
     ]
    }
   ],
   "source": [
    "# Instantiate the model\n",
    "model = LogisticRegression()\n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "estimators = [('normalise', StandardScaler()),\n",
    "              ('model', LogisticRegression())]\n",
    "\n",
    "pipe = Pipeline(estimators)\n",
    "\n",
    "param_grid = [\n",
    "            {'model': [LogisticRegression()], \n",
    "             'normalise': [StandardScaler(), None],\n",
    "             'model__penalty': ['l1', 'l2'],\n",
    "             'model__solver': ['saga', 'liblinear'],\n",
    "             'model__C': [0.001, 0.01, 0.1, 1, 10, 15]}]\n",
    "\n",
    "grid = GridSearchCV(pipe, param_grid, cv=5)\n",
    "fittedgrid = grid.fit(X_train, y_train)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "capstone_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
